### PR TITLE
Adapt Project resource according to latest decisions

### LINKED ISSUE #357: Dedicated Member/Admin resources
Now that we have [dedicated `Project` resources](https://github.com/gardener/gardener/blob/master/example/05-project-dev.yaml) we also want to get rid of the implicit/unversioned logic of adding members to projects and administrators to the whole Garden cluster.

Currently, Gardener relies on the [`garden.sapcloud.io:system:project-member` `ClusterRole`](https://github.com/gardener/gardener/blob/master/charts/gardener/templates/rbac.yaml#L161-L210) in the Garden cluster.
* A project member is any `User`/`Group`/`ServiceAccount` that is part of a `RoleBinding` named `garden-project-members` that was labelled with `garden.sapcloud.io/role=members` and binds to the mentioned `ClusterRole` (the `RoleBinding` must have exactly this name and label so that the Gardener dashboard can understand and work with it).
* An administrator is any `User`/`Group`/`ServiceAccount` that is part of a `ClusterRoleBinding` named `garden-administrators` that was labelled with `garden.sapcloud.io/role=admins` and binds to the mentioned `ClusterRole` (again, same conditions apply due to the Gardener dashboard).

Both ways are hard to understand, too implicit and unversioned. We want to have a better way of binding members to projects and defining administrators.

One approach could be to introduce dedicated `Member` or `Admin` resources in the Garden API, together with corresponding controllers which would be responsible for the underlying RBAC logic. It could look like that:

```yaml
apiVersion: garden.sapcloud.io/v1beta1
kind: Member
metadata:
  name: john-doe
  namespace: garden-my-project
spec:
  subject:
    apiGroup: rbac.authorization.k8s.io
    kind: User
    name: john.doe@example.com
  project: my-project
---
apiVersion: garden.sapcloud.io/v1beta1
kind: Admin
metadata:
  name: alice
spec:
  subject:
    apiGroup: rbac.authorization.k8s.io
    kind: User
    name: john.doe@example.com
```

Another approach could be to include the list of members into the now existing `Project` resources, e.g. like that:

```yaml
apiVersion: garden.sapcloud.io/v1beta1
kind: Project
metadata:
  name: my-project
spec:
  owner:
    apiGroup: rbac.authorization.k8s.io
    kind: User
    name: john.doe@example.com
# description: "This is my first project"
# purpose: "Experimenting with Gardener"
  members:
  - apiGroup: rbac.authorization.k8s.io
    kind: User
    name: john.doe@example.com
  - apiGroup: rbac.authorization.k8s.io
    kind: User
    name: alice@example.com
```

However, how would be administrators realised? We would still need a dedicated resource like in the first proposal.

Both approaches would be easier to implement a) for the Gardener dashboard or b) for any client that wants to talk to the Gardener.

Do we have other ideas or general comments/suggestions?

/cc @holgerkoser @petersutter @grolu @mvladev

### LINKED ISSUE #409: Namespace deletions may lead to projects without corresponding namespaces
Sometimes the deletion of a namespace that a project references goes unnoticed in our project controller. We have to add adapt our code / add a garbage collection loop that can collect and delete such orphaned projects.

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
We have added the `Project` resource to the Garden API with #330. However, it was not possible to manage the members of a project directly with the resource. Moreover, the implementation was violating some best-practices and conventions of Kubernetes (e.g., an admission controller created resources, effectively causing side-effects).

**Which issue(s) this PR fixes**:
Fixes #357
Fixes #409

**Special notes for your reviewer**:
With this change we
1. add a new optional field `.spec.createdBy` (`*rbacv1.Subject`, immutable).
1. add a new optional field `.spec.members` (`[]rbacv1.Subject`, resulting in `RoleBinding`s in the corresponding `Namespace` that binds the members to the "project-member" `ClusterRole` (this was done by the Dashboard before)).
1. change a required field to an optional field `.spec.owner` (`*rbacv1.Subject`).
1. do no longer generate conditions for `Project`s (`.status.conditions` removed).
1. do generate `Events` for `Project`s and write its state into `.status.phase` (one of `Pending`, `Ready`, `Terminating`, `Failed`).
1. remove the introduced admission controller that was creating `Namespace`s for `Project`s.
1. move all responsibilities/logic for creating, reconciling, and deleting `Project`s to the `Project` controller (we are now leveraging Kubernetes' optimistic concurrency controls to prevent race conditions).
1. remove the migration loop that was automatically creating `Project` resources out of properly labelled `Namespace`s before starting the "shoot" controller.
1. remove the "project-namespace controller" that was automatically creating `Project` resources out of properly labelled `Namespace`s (during runtime).
1. migrate all existing `RoleBinding`s created by prior versions of Gardener dashboard, i.e. we add all the subjects to the new `.spec.members` field of the corresponding `Project` (after that we delete these legacy `RoleBinding`s).

:warning: The Gardener dashboard is finalizing the [adaptation](https://github.com/gardener/dashboard/tree/project) of these changes (gardener/dashboard#231). Hence, once this is merged the next Gardener version won't work with dashboard versions < 1.23.

PS: I have prefixed the `api-versions` and `tls-cipher-suites` charts with `utils-` and adapted all remaining charts. /cc @mvladev 

**Release note**:
```noteworthy operator
The `Project` resource has been changed to support management of project members (by setting `.spec.members`, please take a look at the [example `Project` resource](https://github.com/gardener/gardener/blob/master/example/05-project-dev.yaml)).
```
```action operator
:warning: The migration logic for automatically creating `Project` resource out of properly labelled `Namespace`s has been removed. Upgrading from a Gardener version prior 0.11.0 is not supported. Also, this Gardener version is not compatible with Gardener Dashboard versions prior 1.23.0.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -137,3 +137,8 @@ test-cov:
 test-clean:
 	@find . -name "*.coverprofile" -type f -delete
 	@rm -f gardener.coverage.html
+
+.PHONY: generate
+generate:
+	@./hack/generate-code
+	@./hack/generate-examples
```
Comment (zanetworker) [inline Makefile:144]: Maybe separate them into two steps for two reasons: 
- if I just want to execute one independently
- if one fails we don't need to execute the other 

```makefile
generate-code:
     @./hack/generate-code

generate-examples: generate-code 
     @./hack/generate-examples
```

File: pkg/controller/project/project_control_delete.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	// To prevent "stealing" namespaces by other projects we only delete the namespace if its labels match
+	// the project labels.
+	if !apiequality.Semantic.DeepDerivative(namespaceLabelsFromProject(project), namespace.Labels) {
+		return true, nil
+	}
+
+	// If the namespace has been already marked for deletion we do not need to do it again.
+	if namespace.DeletionTimestamp != nil {
```
Comment (timuthy) [inline pkg/controller/project/project_control_delete.go]: I'd prefer to do this check before checking for label equality.

File: pkg/apis/garden/v1beta1/defaults.go [hunk 3]
```diff
@@ -179,6 +179,17 @@ func SetDefaults_Seed(obj *Seed) {
 
 // SetDefaults_Project sets default values for Project objects.
 func SetDefaults_Project(obj *Project) {
+	if len(obj.Spec.CreatedBy.APIGroup) == 0 {
+		switch obj.Spec.Owner.Kind {
+		case rbacv1.ServiceAccountKind:
+			obj.Spec.Owner.APIGroup = ""
+		case rbacv1.UserKind:
+			obj.Spec.Owner.APIGroup = rbacv1.GroupName
+		case rbacv1.GroupKind:
```
Comment (zanetworker) [inline pkg/apis/garden/v1beta1/defaults.go]: you can use a multi-value switch here: 

```go 
	switch obj.Spec.Owner.Kind {
		case rbacv1.ServiceAccountKind:
			obj.Spec.Owner.APIGroup = ""
		case rbacv1.UserKind, rbacv1.GroupKind:
                       obj.Spec.Owner.APIGroup = rbacv1.GroupName
```

File: pkg/apis/garden/v1beta1/types.go [hunk 4]
```diff
@@ -406,6 +406,9 @@ type ProjectList struct {
 
 // ProjectSpec is the specification of a Project.
 type ProjectSpec struct {
+	// CreatedBy is a subject representing a user name, an email address, or any other identifier of a user
+	// who created the project.
+	CreatedBy rbacv1.Subject `json:"createdBy,omitempty"`
```
Comment (zanetworker) [inline pkg/apis/garden/v1beta1/types.go]: are you missing an +optional here?

File: pkg/apis/garden/v1beta1/types.go [hunk 5]
```diff
@@ -415,6 +418,9 @@ type ProjectSpec struct {
 	// Purpose is a human-readable explanation of the project's purpose.
 	// +optional
 	Purpose *string `json:"purpose,omitempty"`
+	// Members is a list of subjects representing a user name, an email address, or any other identifier of a user
+	// that should be part of this project.
+	Members []rbacv1.Subject `json:"members,omitempty"`
```
Comment (zanetworker) [inline pkg/apis/garden/v1beta1/types.go:431]: here as well?

File: pkg/apis/garden/validation/validation.go [hunk 6]
```diff
@@ -617,7 +617,10 @@ func ValidateProjectUpdate(newProject, oldProject *garden.Project) field.ErrorLi
 	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&newProject.ObjectMeta, &oldProject.ObjectMeta, field.NewPath("metadata"))...)
 	allErrs = append(allErrs, ValidateProject(newProject)...)
 
-	if !apiequality.Semantic.DeepEqual(newProject.Spec.Namespace, oldProject.Spec.Namespace) {
+	if !apiequality.Semantic.DeepEqual(newProject.Spec.CreatedBy, oldProject.Spec.CreatedBy) {
+		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newProject.Spec.CreatedBy, oldProject.Spec.CreatedBy, field.NewPath("spec", "createdBy"))...)
+	}
+	if oldProject.Spec.Namespace != nil && !apiequality.Semantic.DeepEqual(newProject.Spec.Namespace, oldProject.Spec.Namespace) {
 		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newProject.Spec.Namespace, oldProject.Spec.Namespace, field.NewPath("spec", "namespace"))...)
```
Comment (zanetworker) [inline pkg/apis/garden/validation/validation.go:624]: same check in two places?  maybe just use the check in ValidateImmutableFied(..) and remove the `&&` ?

File: charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-admin.yaml [hunk 7]
```diff
@@ -8,11 +8,11 @@ metadata:
     kind: Project
     blockOwnerDeletion: false
     controller: true
-    name: {{ .Values.project.name }}
-    uid: {{ .Values.project.uid }}
+    name: "{{ .Values.project.name }}"
```
Comment (zanetworker) [inline charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-admin.yaml]: use the  ```{{ .Values.project.name |  quote }}``` ?

File: charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-admin.yaml [hunk 8]
```diff
@@ -8,11 +8,11 @@ metadata:
...
+    uid: "{{ .Values.project.uid }}"
```
Comment (zanetworker) [inline charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-admin.yaml]: use the {{ .Values.project.name | quote }} ?

File: charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-members.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRoleBinding
+metadata:
+  name: garden.sapcloud.io:system:project-member:{{ .Values.project.name }}
+  ownerReferences:
+  - apiVersion: garden.sapcloud.io/v1beta1
+    kind: Project
+    blockOwnerDeletion: false
+    controller: true
+    name: "{{ .Values.project.name }}"
```
Comment (zanetworker) [inline charts/garden-project/charts/project-rbac/templates/clusterrolebinding-project-members.yaml]: same, use `quote` function?

File: pkg/apis/garden/validation/validation.go [hunk 10]
```diff
@@ -617,9 +617,15 @@ func ValidateProjectUpdate(newProject, oldProject *garden.Project) field.ErrorLi
...
 
-	if !apiequality.Semantic.DeepEqual(newProject.Spec.Namespace, oldProject.Spec.Namespace) {
+	if oldProject.Spec.CreatedBy != nil && !apiequality.Semantic.DeepEqual(newProject.Spec.CreatedBy, oldProject.Spec.CreatedBy) {
+		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newProject.Spec.CreatedBy, oldProject.Spec.CreatedBy, field.NewPath("spec", "createdBy"))...)
+	}
+	if oldProject.Spec.Namespace != nil && !apiequality.Semantic.DeepEqual(newProject.Spec.Namespace, oldProject.Spec.Namespace) {
 		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newProject.Spec.Namespace, oldProject.Spec.Namespace, field.NewPath("spec", "namespace"))...)
 	}
+	if oldProject.Spec.Owner != nil && newProject.Spec.Owner == nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "owner"), newProject.Spec.Owner, "owner cannot be reset"))
```
Comment (zanetworker) [inline pkg/apis/garden/validation/validation.go:627]: Maybe use the same message all over for immutable fields, for example "owner field is immutable"?

File: pkg/apis/garden/validation/validation.go [hunk 11]
```diff
@@ -677,6 +691,10 @@ func ValidateSubject(subject rbacv1.Subject, fldPath *field.Path) field.ErrorLis
 func ValidateProjectStatusUpdate(newProject, oldProject *garden.Project) field.ErrorList {
 	allErrs := field.ErrorList{}
 
+	if len(oldProject.Status.Phase) > 0 && len(newProject.Status.Phase) == 0 {
```
Comment (zanetworker) [inline pkg/apis/garden/validation/validation.go:694]: maybe use "phase cannot be s/set/updated/g to an empty string"

File: Makefile (hunk 1 above)
Comment (rfranzke) [reply Makefile:144]: It's safer to have one command that generates both to keep the results in sync (e.g., if the API and the examples have changed but the generation was not yet run).
If only one generation may be executed then  just `./hack/generate-{code,examples}` can be called.

File: pkg/controller/project/project_control_delete.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"github.com/sirupsen/logrus"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func (c *defaultControl) delete(project *gardenv1beta1.Project, projectLogger logrus.FieldLogger) (bool, error) {
```
Comment (zanetworker) [inline pkg/controller/project/project_control_delete.go:26]: we  can be consistent and put comments on all private functions?

File: pkg/apis/garden/validation/validation.go (hunk 6 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:624]: Do you suggest

```golang
if oldProject.Spec.Namespace != nil {
  allErrs = append(allErrs, apivalidation.ValidateImmutableField(newProject.Spec.Namespace, oldProject.Spec.Namespace, field.NewPath("spec", "namespace"))...)
}
```
?

I guess this makes sense, thank you, will do.

File: pkg/apis/garden/v1beta1/defaults.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/garden/v1beta1/defaults.go]: I've copied the validation from Kubernetes itself. Let's keep it as it is, this way it's also easier to read.

File: pkg/controller/project/project_control_reconcile.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+	// If the project has no phase yet then we update it to be 'pending'.
+	if len(project.Status.Phase) == 0 {
+		if _, err := c.updateProjectStatus(project.ObjectMeta, setProjectPhase(gardenv1beta1.ProjectPending)); err != nil {
+			return err
+		}
+	}
+
+	// We reconcile the namespace for the project: If the .spec.namespace is set then we try to claim it, if it is not
+	// set then we create a new namespace with a random hash value.
+	namespace, err := c.reconcileNamespace(project)
```
Comment (zanetworker) [inline pkg/controller/project/project_control_reconcile.go]: reconcileNamespaceForProject()?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/validation/validation.go (hunk 10 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:627]: But it is mutable. It's just not allowed to "reset" it, i.e. setting it to `nil`.

File: pkg/controller/project/project_control_delete.go (hunk 12 above)
Comment (rfranzke) [reply pkg/controller/project/project_control_delete.go:26]: What for? It's clear what this function does, isn't it?

**Comment by rfranzke** [comment]
Thanks for the feedback, I've addressed it and rebased the branch.

**Review by timuthy — DISMISSED** [review]

File: pkg/controller/project/project.go [hunk 14]
```diff
@@ -132,12 +125,19 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 
 	logger.Logger.Info("Project controller initialized.")
 
+	// Migration of legacy rolebinding to new .spec.members field in project resource
+	go func() {
+		utils.Retry(10*time.Second, 10*time.Minute, func() (ok, severe bool, err error) {
```
Comment (adracus) [inline pkg/controller/project/project.go:133]: Move logging of the project migration function to here and log only once.

File: pkg/controller/project/project.go [hunk 15]
```diff
@@ -159,3 +159,47 @@ func (c *Controller) Run(ctx context.Context, workers int) {
...
 	return c.numberOfRunningWorkers
 }
+
+// This function has been introduced to migrate the legacy project member rolebindings created by older
+// versions of the Gardener dashboard (prior 1.23) to the new .spec.members field of the Project resource.
+// It can be removed in future versions of Gardener.
+func (c *Controller) migrateRoleBindingToProjectMembers() error {
+	projectList, err := c.projectLister.List(labels.Everything())
+	if err != nil {
+		logger.Logger.Errorf("[ROLEBINDING MIGRATION] Error fetching projects: %+v", err)
```
Comment (adracus) [inline pkg/controller/project/project.go]: Remove logging / use error wrapping to have a pretty message (`go/x/errors` package).

File: pkg/controller/project/project.go [hunk 16]
```diff
@@ -159,3 +159,47 @@ func (c *Controller) Run(ctx context.Context, workers int) {
...
+	}
+
+	for _, roleBinding := range roleBindingList.Items {
+		if projectName, ok := namespaceToProject[roleBinding.Namespace]; ok {
+			if _, err := kutils.TryUpdateProject(c.k8sGardenClient.GardenClientset(), retry.DefaultBackoff, metav1.ObjectMeta{Name: projectName}, func(project *gardenv1beta1.Project) (*gardenv1beta1.Project, error) {
+				project.Spec.Members = roleBinding.Subjects
+				return project, nil
+			}); err != nil {
+				logger.Logger.Errorf("[ROLEBINDING MIGRATION] Error updating the project members based on the legacy rolebinding: %+v", err)
+				continue
```
Comment (adracus) [inline pkg/controller/project/project.go:201]: Maybe accumulate the errors in a `multierror` and return the accumulated error in the end?

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 17]
```diff
@@ -119,24 +129,70 @@ func (d *DeletionConfirmation) Admit(a admission.Attributes) error {
...
+	)
 
 	switch a.GetKind().GroupKind() {
 	case garden.Kind("Shoot"):
-		shoot, err := d.shootLister.Shoots(a.GetNamespace()).Get(a.GetName())
-		if err != nil {
-			return err
+		cacheLookup = func() (metav1.Object, error) {
+			shoot, err := d.shootLister.Shoots(a.GetNamespace()).Get(a.GetName())
+			if err != nil {
```
Comment (adracus) [inline plugin/pkg/global/deletionconfirmation/admission.go]: why not just immediately return the value of the function (the 'cast' should happen implicitly)

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 18]
```diff
@@ -119,24 +129,70 @@ func (d *DeletionConfirmation) Admit(a admission.Attributes) error {
...
+				return nil, err
+			}
+			return shoot, nil
 		}
-		obj = shoot
+		liveLookup = func() (metav1.Object, error) {
+			shoot, err := d.gardenClient.Garden().Shoots(a.GetNamespace()).Get(a.GetName(), metav1.GetOptions{})
```
Comment (adracus) [inline plugin/pkg/global/deletionconfirmation/admission.go]: Same as above

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 19]
```diff
@@ -119,24 +129,70 @@ func (d *DeletionConfirmation) Admit(a admission.Attributes) error {
...
+			if err != nil {
+				return nil, err
+			}
+			return project, nil
 		}
-		obj = project
 	}
 
-	annotations := obj.GetAnnotations()
+	if cacheLookup == nil || liveLookup == nil {
```
Comment (adracus) [inline plugin/pkg/global/deletionconfirmation/admission.go]: Use default case instead to indicate that the kind is unknown and also print it

**Pull request merged**